netdata/netdata
# Final migration of release code off of Travis CI.

##### Summary

This finishes migrating the release process from Travis CI to GitHub Actions.

Once this is merged, instead of triggering releases by pushing a magic empty commit with a special commit message, they will be triggered by manually triggering the release workflow from the GitHub UI with the correct version number and release type information.

Additionally, nightly builds will be run via the same workflow as a scheduled run.

##### Test Plan

The core logic for handling the release preparation is all in the `.github/scripts/prepare_changelog.sh` script.  Testing this script requires an isolated environment with no `git` configuration done ahead of time that includes a working Docker install, and a GitHub Personal Access Token (both Docker and the PAT are required by the changelog generation tool we use).

1. Clone the `netdata/netdata` repo.
2. Explicitly remove the `origin` remote with `git remote remove origin`. This should not be required, but is extra insurance just in case the script does not properly recognize that it is not running in GitHub Actions.
3. Download the above-mentioned script from this PR.
4. Export an environment variable named `GITHUB_TOKEN` containing the PAT.
5. Run the script like so: `prepare_changelog.sh netdata/netdata <event_name> <type> <version>`.

The `<event_name>` can be anything. The exact value `schedule` has special meaning.
Valid values for `<type>` are `nightly`, `patch`, `minor`, and `major`, anything else should result in an error message.
Valid values for `<version>` are either `nightly` or a semver-style version number starting with a `v` and containing exactly three components consisting of one or more digits in the range 0-9. Anything else should result in an error message.

Expected behavior is as follows:

* Supplying any value other than `netdata/netdata` as the first argument should result in an exit code of zero and three lines of output (one notice message, and two `set-output` commands setting the values of the `run` and `version` outputs).
* Specifying an event name of `schedule` should produce output messages about preparing a nightly release, and should result in a single commit being added to the master branch modifying `packaging/version` and `CHANGELOG.md` with a title talking about nightly releases. This should ignore the type and version values.
* Specifying any other event name together with a type of `nightly` should produce identical results to specifying an event name of `schedule`.
* Specifying any other event name together with a type of `major` should validate the version (see below for info on what all this entails) and then create a new commit on the master branch modifying `packaging/version` and `CHANGELOG.md` and mentioning it as a major release, and also create a new tag with a name matching the version number passed in pointing to that commit.
* Specifying any other event name together with a type of `minor` should validate the version, validate that a release branch for that version does not already exist, create a new branch with a name derived from the version with the patch number removed (for example, for version v1.35.0, the branch name would be v1.35), and a commit to that branch modifying `packaging/version` and `CHANGELOG.md` and mentioning it as a minor release, and also create a new tag with a name matching the version number pointing to that commit.
* Specifying any other event name together with a type of `patch` should validate the version, confirm that a release branch for the associated minor version exists in the repo, create a commit on that release branch modifying `packaging/version` and `CHANGELOG.md` and mentioning it as a patch release, and also create a new tag with a name matching the version number pointing to that commit.
* Specifying any other event name together with a type not listed above, or specifying a version of `nightly` with a type of `major`, `minor`, or `patch` should report an error and exit with a non-zero exit status.

Version number validation checks all the following, reporting an error and exiting with a non-zero exit status if any of them fail:
* The version number matches the correct format (exact check is against the POSIX ERE `^v[[::digit::]]+\.[[::digit::]]+\.[[::digit::]]+$`).
* There is no tag with a matching name already present in the repository.
* The superior version numbers (major for minor releases and major and min0r for patch releases) match what is already in the repo (for example, trying to use 2.33.2 for a patch release when the version in the repo is 1.33.1).
* The inferior version numbers (patch for major and minor and minor for major) are exactly zero (for example, trying to use a version of 2.0.1 for a major release).
* The version component associated with the release type is strictly greater than what is currently in the repo, though not necessarily exactly sequentially larger (for example, trying to release 1.33.0 when the version in the repo is 1.33.1 should fail, but 1.33.3 would work in the same case).

The workflow itself cannot be readily tested in this PR and will need to be validated through code inspection and static checking.

##### Additional Information

This also brings release-branch support for minor and patch releases, meaning that we will not need the extended soft code freezes starting with the first (successful) release after this gets merged.

Release branch support for major releases is not included due to the complexity involved. We can look into that if desired at a later date when we know we will actually be doing a major release.

Final removal of the Travis CI config and scripts will be done as a separate PR, as that will require special synchronization with changes on the Travis CI website to avoid breaking CI.
